strawberry/schema/subscribe.py

```diff
@@ -0,0 +1,166 @@
+                    PreExecutionError(data=None, errors=aiter_or_result.errors),
+                    extensions_runner,
+                    process_errors,
+                )
+            else:
+                aiterator = aiter_or_result.__aiter__()
+                running = True
+                while running:
+                    # reset extensions results for each iteration
+                    execution_context.extensions_results = {}
+                    async with extensions_runner.executing():
+                        try:
+                            origin_result: Union[
+                                ExecutionResult, OriginalExecutionResult
+                            ] = await aiterator.__anext__()
+
+                        except StopAsyncIteration:
+                            break
+                        except Exception as exc:
+                            # graphql-core doesn't handle exceptions raised in the async generator.
+                            origin_result = ExecutionResult(
+                                data=None, errors=[_coerce_error(exc)]
+                            )
+                            running = False
+                    # we could have yielded in the except block above.
+                    # but this way we make sure `get_result` hook is called deterministically after
+                    # `on_execute` hook is done.
+                    yield await _handle_execution_result(
+                        execution_context,
+                        origin_result,
+                        extensions_runner,
+                        process_errors,
+                    )
+        # catch exceptions raised in `on_execute` hook.
+        except Exception as exc:
+            origin_result = OriginalExecutionResult(
+                data=None, errors=[_coerce_error(exc)]
+            )
+            yield await _handle_execution_result(
+                execution_context,
+                origin_result,
+                extensions_runner,
+                process_errors,
+            )
+
+
+async def subscribe(
+    schema: GraphQLSchema,
+    execution_context: ExecutionContext,
+    extensions_runner: SchemaExtensionsRunner,
+    process_errors: ProcessErrors,
+    middleware_manager: MiddlewareManager,
+    execution_context_class: Optional[Type[GraphQLExecutionContext]] = None,
+) -> SubscriptionResult:
+    asyncgen = _subscribe(
+        schema,
+        execution_context,
+        extensions_runner,
+        process_errors,
+        middleware_manager,
+        execution_context_class,
+    )
+    # GrapQL-core might return an initial error result instead of an async iterator.
+    # This happens when "there was an immediate error" i.e resolver is not an async iterator.
+    # To overcome this while maintaining the extension contexts we do this trick.
+    first = await asyncgen.__anext__()
+    if isinstance(first, PreExecutionError):
+        await asyncgen.aclose()
```

Comment:
couldn't we just check if there's an error to the response, isn't that enough?

Reply:
hmm might be. though this is more explicit IMO. Websockets integrations rely on this on protocol level.